File: leetcode/561/v2.cpp
class Solution {
public:
    int arrayPairSum(vector<int>& nums) {
        // BucketSort
        
        const int maxNumAllowed = 10000;
        const int arrSize = 2 * maxNumAllowed + 1;
        // The numbers range from -10k to +10k, so 20k
        vector<int> frequency (arrSize, 0);
        
        // Fill frequency array
        for (auto x : nums)
            frequency[x + maxNumAllowed]++;
         
        
        // Iterating the whole frequency array, to find
        bool shallSum = true;
        int sum {};
        for (int i{}; i < arrSize; i++) {
            while (frequency[i] > 0) {
                if (shallSum) {
                    // The index i doesn't represent the number, it represents
                    // i = num + 10000  ->  num = i - 10000
                    sum += i - maxNumAllowed; 
                    shallSum = false;
                } else {
                    shallSum = true;
                }
                // If i don't decrement, the while will loop forever
                frequency[i]--;
            }
        }
        return sum;
        
        
    }
};

File: leetcode/238/v2.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix = [0 for _ in range(len(nums))]
        postfix = [0 for _ in range(len(nums))]

        prefix[0] = nums[0]
        postfix[-1] = nums[-1]

        for i in range(1, len(nums)):
            prefix[i] = prefix[i - 1] * nums[i]

        for i in range(len(nums) - 2, -1, -1):
            postfix[i] = postfix[i + 1] * nums[i]

        ans = [0 for _ in range(len(nums))]
        ans[0] = postfix[1]
        ans[-1] = prefix[-2]
        for i in range(1, len(nums) - 1):
            ans[i] = (prefix[i - 1] * postfix[i + 1])
        
        return ans

